eessi/testsuite/hooks.py

```diff
@@ -282,16 +282,50 @@ def _assign_one_task_per_gpu(test: rfm.RegressionTest):
     log(f'num_cpus_per_task set to {test.num_cpus_per_task}')
     log(f'num_tasks set to {test.num_tasks}')
 
+def _set_or_append_valid_systems(test: rfm.RegressionTest, valid_systems: str):
+    """
+    Sets test.valid_systems based on the valid_systems argument.
+    - When valid_systems is an empty string, test.valid_systems will be set equal to eessi.testsuite.constants.INVALID_SYSTEM
+    - When no test.valid_system was set yet, or it was at the default value ['*'], it will be overwritten by [valid_system]
+    - When a test.valid_system was already set, valid_system will be appended to it. This allows adding requests for multiple partition features by different hooks.
+    """
+
+    # This indicates an invalid test that always has to be filtered
+    if valid_systems == '':
+        test.valid_systems = [INVALID_SYSTEM]
+        return
+
+    # test.valid_systems wasn't set yet, so set it
+    if len(test.valid_systems) == 0:
+        test.valid_systems = [valid_systems]
+    # test.valid_systems still at default value, so overwrite
+    elif len(test.valid_systems) == 1 and test.valid_systems[0] == '*':  
+        test.valid_systems = [valid_systems]
```

Comment:
shouldn't we add a condition that checks if `len(test.valid_systems) > 1`?
the hooks don't support multiple items in the list, so it may be safer to error out with a clear message if the user has set it on the command line or in a custom test.

in any case, we can remove the first equality of the above condition: `len(test.valid_systems) == 1`, because, if the first item is `'*'`, the test will always run (not filtered) regardless of what the other items in the list are.

Reply:
The `len(test.valid_systems) == 1` was to protect `test.valid_systems[0] == '*'` from failing: we are accessing the first element there. If the thing contains zero elements, accessing `test.valid_systems[0]` will cause a hard error. It's not strictly needed since we already checked for `if len(test.valid_systems) == 0:` I guess. In the end, it depends on whether we keep that or not - lets decide that first. Actually, I would probably start with `if len(test.valid_systems) == 0:` regardless, and throw a specific `ValueError` for that case (namely that the default for `valid_systems` in a test class should be `['*']`, if we indeed want to enforce that).

Reply:
Ok, I decided to keep the `len(test.valid_systems) == 1` test. If it's longer than 1, I want it to go to the `else` statement and say "I don't know what to do here". That is the only reasoanable thing to do, because what should I overwrite if someone somehow had `test.valid_systems = [ '*', 'my_system_name']`? (it's an unlikely value to have, but it is not impossible)